lib/datastore/index.js

```diff
@@ -1,132 +1,260 @@
+   * @param {Corestore} options.corestore
+   * @param {import('../sqlite.js').Sqlite} options.sqlite an instance of the internal Sqlite class
+   *
    */
-  constructor({ dataType, corestore, indexer }) {
-    this.#dataType = dataType
+  constructor(options) {
+    const { identityPublicKey, keyPair, corestore, sqlite } = options
+
+    this.#identityPublicKey = identityPublicKey
+    this.#keyPair = keyPair
     this.#corestore = corestore
-    this.#indexer = indexer
-    this.#writer = this.#corestore.get({ name: 'writer' })
+    this.#sqlite = sqlite
+
+    this.#indexer = new MultiCoreIndexer(this.cores, {
+      /**
+       * @param {String} key
+       */
+      storage: (key) => {
+        return new ram(key)
+      },
+      /**
+       * @param {Entry[]} entries
+       */
+      batch: (entries) => {
+        for (const entry of entries) {
+          const { block } = entry
+
+          const dataType = this.getDataTypeForBlock(block)
+
+          if (!dataType) continue
+
+          dataType.index([block])
+        }
+      },
+    })
+  }
+
+  get localCore() {
+    return this.#writer
+  }
+
+  get cores() {
+    return [...this.#corestore.cores.values()]
+  }
+
+  get dataTypes() {
+    return [...this.#dataTypes.values()]
   }
 
-  /**
-   * Wait for the corestore and writer hypercore to be ready
-   * @returns {Promise<void>}
-   */
   async ready() {
-    await this.#writer.ready()
     await this.#corestore.ready()
-  }
+    const secretKey = this.#keyPair.secretKey
+    const publicKey = this.#keyPair.publicKey
 
-  /**
-   * Validate a doc
-   * @param {Doc} doc
-   * @returns {Boolean | PromiseLike<Boolean>}
-   * @throws {Error}
-   */
-  validate(doc) {
-    return this.#dataType.validate(doc)
+    this.#writer = this.#corestore.get({
+      keyPair: this.#keyPair,
+      sparse: false,
+      auth: {
+        /**
+         * @param {Buffer} signable
+         */
+        sign(signable) {
+          const signature = b4a.alloc(sodium.crypto_sign_BYTES)
+          sodium.crypto_sign_detached(signature, signable, secretKey)
+          return signature
+        },
+        /**
+         * @param {Buffer} signable
+         * @param {Buffer} signature
+         */
+        verify(signable, signature) {
+          return sodium.crypto_sign_verify_detached(
+            signature,
+            signable,
+            publicKey
+          )
+        },
+      },
+    })
+
+    await this.#writer.ready()
+    await this.#indexer.addCore(this.#writer)
   }
 
   /**
-   * Encode a doc (an object), to a block (a Buffer)
-   * @param {Doc} doc
-   * @returns {Block}
+   * Create a new DataType in the DataStore
+   * @param {Object} options
+   * @param {String} options.name
+   * @param {Object} options.schema
+   * @param {BlockPrefix} options.blockPrefix
+   * @param {String} options.extraColumns
+   * @returns
    */
-  encode(doc) {
-    return this.#dataType.encode(doc)
+  dataType(options) {
```

Comment:
from @gmaclennan:

> Also wondering about the way that datatypes can be added after creating the datastore instance. We don't restart indexing when adding a new datatype, so as I understand it this can result in data not being indexed, e.g. create DataStore instance on cores with existing data -> starts indexing -> a few ticks later call dataStore.dataType() to add a data type -> items indexed by multi-core indexer will not be included in the index.

> I'm wondering if the datatypes should just be in the constructor to simplify things? Also it seems at first glance that cores would be added multiple times to the multicore indexer, because they are added on initialization of the datastore, then from what I understand the same cores are re-added for each datatype that is added, since we basically one core for each device in the project.